### PR TITLE
Update golang version to 1.17

### PR DESCRIPTION
**Pull request by elmiko**
This change updates the go.mod, makefile, dockerfiles, and the vendor directory to use golang 1.17 for building. it also adds a minor convenience option to the makefile that allows a user to override the build image. It also adds a fix to the Makefile for building with containers.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -57,10 +58,10 @@ ifeq ($(USE_DOCKER), 1)
 endif
 
 ifeq ($(NO_DOCKER), 1)
-  DOCKER_CMD =
+  DOCKER_CMD = CGO_ENABLED=0 &&
```
Comment (JoelSpeed) [inline Makefile:61]: This doesn't appear to actually do anything, we already have cgo disabled by line 39 don't we?

What's the `&&` for in this case?

File: go.mod [hunk 2]
```diff
@@ -21,3 +21,89 @@ require (
 	sigs.k8s.io/controller-runtime v0.9.6
 	sigs.k8s.io/controller-tools v0.6.3-0.20210916130746-94401651a6c3
 )
+
+require (
```
Comment (JoelSpeed) [inline go.mod:25]: Does this need a `go mod tidy`? Or is this a new go 1.17 thing?

File: Makefile (hunk 1 above)
Comment (elmiko) [reply Makefile:61]: that's a good point about line 39, i can probably remove this.

the `&&` just says "do the thing to the left, and if it fails return an error before doing the thing on the right", i was trying to make sure we had the env variable set for non-container builds. i kept getting an error that way it was expressed before.

Comment (elmiko) [reply Makefile:61]: i ran a couple tests, and fwiw that variable set on L39 does not get propagated to the build environment. that said, i think we don't need it set explicitly, i tried doing a few test builds and they looked the same whether i had it enabled or not. i will remove the part on L61 just to reduce confusion here.

Comment (elmiko) [reply Makefile:61]: i'm doing some further testing and i'm doubting my previous results. when i set up the build container with the proper env variables i get a clean build, but if i remove the `CGO_ENABLED` i get some warnings about the shared libraries. i am starting to think we need all the env variables loaded into `DOCKER_CMD` when `NO_DOCKER=1`. i will do some further testing though to confirm.

**Review by elmiko — COMMENTED** [review]
i'm going to remove the `CGO_ENABLED` from the `DOCKER_CMD` for now, my only concern here is that we negatively impact the image builds.

**Comment by elmiko** [comment]
i see this in the error output for the images test, this is the same error i get when building locally

```
 CGO_ENABLED=0 &&  go build  -o "bin/machine-controller-manager" \
               -ldflags "-X sigs.k8s.io/cluster-api-provider-aws/pkg/version.Raw=v0.2.0-461-g3be87ba -extldflags "-static"" "sigs.k8s.io/cluster-api-provider-aws/cmd/manager"
# sigs.k8s.io/cluster-api-provider-aws/cmd/manager
/usr/lib/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: cannot find -lpthread
/usr/bin/ld: cannot find -ldl
/usr/bin/ld: cannot find -ldl
/usr/bin/ld: cannot find -lc
collect2: error: ld returned 1 exit status 
```

**Comment by JoelSpeed** [comment]
What happens if you remove the `&&`, I think this should work:
```
CGO_ENABLED=0 go build  -o "bin/machine-controller-manager" \
               -ldflags "-X sigs.k8s.io/cluster-api-provider-aws/pkg/version.Raw=v0.2.0-461-g3be87ba -extldflags "-static"" "sigs.k8s.io/cluster-api-provider-aws/cmd/manager"
```
At the moment I don't think you're setting the env var correctly, the && breaks it I think

**Comment by elmiko** [comment]
update
* fixing the makefile more
* fixing up the community dockerfile

**Review by JoelSpeed — COMMENTED** [review]
/approve

**Comment by JoelSpeed** [comment]
There are some code generation issues that will need to be fixed. IIUC go1.17 introduces some new requirements for build tags that we will need to add with our existing code

**Comment by elmiko** [comment]
good call out, i was not aware of that. i'll go back through and run the generators on these PRs

**Comment by elmiko** [comment]
update
* ran `make check` to get all the formatting changes

also, i checked the generated files by re-running `make generate`, but no changes resulted

**Pull request merged**